### ISSUE TITLE
list of types with multiple bases only uses first shared base?

### ISSUE THREAD

**Issue by davidszotten** [python/mypy#8513]
i may be misunderstanding something, but it seems that when making a list of types, each with multiple shared base classes, mypy mistakenly infers this to be a list of the first base only, missing others

```
$ cat mypytest.py
# two different base classes with different attributes
class B1:
    b1 = None


class B2:
    b2 = None


# two classes that subclass both B1 and B2
class S1(B1, B2):
    pass


class S2(B1, B2):
    pass


# a list of the subclasses
types = [S1, S2]

# it seems mypy thinks `types` is List[Type[B1]]
ok = [x.b1 for x in types]
why_not_ok = [x.b2 for x in types]
```

```
$ mypy mypytest.py
mypytest.py:24: error: "Type[B1]" has no attribute "b2"
```

```
$ mypy --version
mypy 0.770+dev.b2edab25b7e4a6131484bb6edf314e2fa078aee7
$ python -V
Python 3.7.3
```

**Comment by SankarChinnaShanmugam** [comment]
As per my understanding it follows as below:

Every class is also a valid type. Any instance of a subclass is also compatible with all superclasses – it follows that every value is compatible with the object type (and incidentally also the Any type, discussed below). Mypy analyzes the bodies of classes to determine which methods and attributes are available in instances. This example uses
subclassing:
class A:
        def f(self) -> int: # Type of self inferred (A)
               return 2
class B(A):
          def f(self) -> int:
                return 3
          def g(self) -> int:
                 return 4

def foo(a: A) -> None:
           print(a.f()) # 3
            a.g() # Error: "A" has no attribute "g"

foo(B()) # OK (B is a subclass of A)

Please correct me if i wrong.

**Comment by msullivan** [comment]
Yeah, that's definitely true. mypy will not be able to figure this one out. Unfortunately there isn't even really a type that we can write down in our type system --- we'd need intersection types so you could express that it is a `List[Intersect[B1, B2]]`. 

We recently landed some changes to do ad-hoc intersections in certain cases but I don't know if that could be scaled up to help here. @Michael0x2a ?
